pkg/controller/clusterpool/collections.go

```diff
@@ -0,0 +1,548 @@
+		}
+	}
+	logger.WithFields(log.Fields{
+		"assignable": len(ret.assignable),
+		"claimed":    len(ret.byClaimName),
+		"deleting":   len(ret.deleting),
+		"installing": len(ret.installing),
+		"unclaimed":  len(ret.installing) + len(ret.assignable),
+	}).Debug("found clusters for ClusterPool")
+	return &ret, nil
+}
+
+// ByName returns the named ClusterDeployment from the cdLookup, or nil if no CD by that name exists.
+func (cds *cdCollection) ByName(cdName string) *hivev1.ClusterDeployment {
+	cd, _ := cds.byCDName[cdName]
+	return cd
+
+}
+
+// Total returns the total number of ClusterDeployments in the cdLookup.
+func (cds *cdCollection) Total() int {
+	return len(cds.byCDName)
+}
+
+// Assignable returns a list of ClusterDeployment refs, sorted according to which we feel ought to
+// be assigned first to minimize claim response time.
+// - Running CDs, oldest first
+// - Resuming CDs, in order of least recently resumed (soonest likely to become Running)
+// - Hibernating CDs, oldest first
+// - Anything else, oldest first. (These should probably be skipped. TODO?)
+func (cds *cdCollection) Assignable() []*hivev1.ClusterDeployment {
+	if !cds.assignableSorted {
+		sort.Slice(
+			cds.assignable,
+			func(i, j int) bool {
+				// Look up the Hibernating condition for both elements
+				iCond := controllerutils.FindClusterDeploymentCondition(cds.assignable[i].Status.Conditions, hivev1.ClusterHibernatingCondition)
+				jCond := controllerutils.FindClusterDeploymentCondition(cds.assignable[j].Status.Conditions, hivev1.ClusterHibernatingCondition)
+				// We should always have a Hibernating condition, but in case we don't, use the
+				// empty string as the "Reason". This will default down below to sorting by
+				// creation timestamp. (That's kind of arbitrary because we shouldn't end up using
+				// such CDs anyway.)
+				var iReason, jReason string
+				if iCond != nil {
+					iReason = iCond.Reason
+				}
+				if jCond != nil {
+					jReason = jCond.Reason
+				}
+				// Prioritize Running => Resuming => Hibernating => anything else
+				switch iReason != jReason {
+				case iReason == hivev1.RunningHibernationReason:
+					return true
+				case jReason == hivev1.RunningHibernationReason:
+					return false
+				case iReason == hivev1.ResumingHibernationReason:
+					return true
+				case jReason == hivev1.ResumingHibernationReason:
+					return false
+				case iReason == hivev1.HibernatingHibernationReason:
+					return true
+				case jReason == hivev1.HibernatingHibernationReason:
+					return false
+					// If neither has a "valid" state, fall through and we'll end up sorting by age
+				}
+				// Special case: if both are Resuming, sort by the least recently resumed -- i.e.
+				// the one likely to become Running the soonest.
+				if iReason == hivev1.ResumingHibernationReason && jReason == hivev1.ResumingHibernationReason {
+					return iCond.LastTransitionTime.Before(&jCond.LastTransitionTime)
+				}
```

Comment:
Previously the list of assignable CDs was unsorted -- it would be in whatever order they came back from the server. This could potentially mean some CDs get really stale (cf. HIVE-1058 again). Beyond that, I was trying to anticipate [hot spares](https://github.com/openshift/hive/pull/1434/files#diff-15704998060bbad90e0abcdb8d4fc87226eac4e7f57719cca93302d88614703dR84) -- but in my most recent update there I decided it should be much simpler than this. I'll change it to just sort by `creationTimestamp`.